src/hotspot/share/jvmci/jvmci.hpp

```diff
@@ -163,6 +163,115 @@ class JVMCI : public AllStatic {
   static void event2(const char* format, ...) ATTRIBUTE_PRINTF(1, 2);
   static void event3(const char* format, ...) ATTRIBUTE_PRINTF(1, 2);
   static void event4(const char* format, ...) ATTRIBUTE_PRINTF(1, 2);
+
+  static jlong* _jvmci_old_thread_counters;
+  static void collect_counters(jlong* array, int length);
+
+  static bool resize_counters(JavaThread* thread, int current_size, int new_size);
+
+  static bool resize_all_jvmci_counters(int new_size);
+
+  static void free_counters(JavaThread* thread);
+
+  static void accumulate_counters(JavaThread* thread);
+
+  static void init_counters();
+
+  static void free_counters();
+};
+
+
+
+class JVMCIThreadState {
+ public:
+  JVMCIThreadState();
+
+ private:
+  friend class JavaThread;
+  friend class JVMCI;
+  friend class JVMCIVMStructs;
+
+  // The _pending_* fields below are used to communicate extra information
+  // from an uncommon trap in JVMCI compiled code to the uncommon trap handler.
+
+  // Communicates the DeoptReason and DeoptAction of the uncommon trap
+  int       _pending_deoptimization;
```

Comment:
Nit: Why the extra large alignment spacing of all the declarations? (I'm not a fan of such alignment as it is too hard to maintain - and too hard to type in the first place!)

Reply:
That probably comes from a time before we nicely commented each JVMCI field ;-)
I agree that there's no need for the alignment now.

Reply:
I removed the extra indenting.